BUG: TimedeltaIndex raising ValueError when slice indexing (#16637)

jdeschenes opened the pull request:
Before accepting this Pull request is there any other kind of usage of loc that might fail using deltatimeindex that I am not aware of? I could take some time to fix it once and for all.

closes #16637

 - [ x ] closes #16637
 - [ x ] tests added / passed
 - [ x ] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [ x ] whatsnew entry

Commit:
BUG: TimedeltaIndex raising ValueError when slice indexing (#16637)

closes #16637

Commit:
v0.20.2 is already released

jreback commented:
https://github.com/pandas-dev/pandas/issues?q=is%3Aopen+is%3Aissue+label%3ATimedelta+label%3AIndexing

are the only timedelta / indexing issues AFAICT (aside from new  one)

jreback commented on pandas/tests/indexing/test_timedelta.py:
```diff
@@ -18,3 +18,20 @@ def test_boolean_indexing(self):
                                     index=pd.to_timedelta(range(10), unit='s'),
                                     columns=['x'])
             tm.assert_frame_equal(expected, result)
+
```
just make sure test is near other slicing testsing (may not actually be many).

jreback commented:
lgtm (just a comment to check location of test). ping on green.

Commit:
Fixed unit tests.

Commit:
This test would fail on 32 bits.

jorisvandenbossche commented on pandas/tests/indexing/test_timedelta.py:
```diff
@@ -13,7 +13,24 @@ def test_boolean_indexing(self):
+                                    columns=['x'])
+            tm.assert_frame_equal(expected, result)
+
+    def test_slice_indexing(self):
```
this is not slice indexing, but list-indexing with another index, if I am correct ? 
If so, I would update the name (and also whatsnew file)

jdeschenes replied:
It was meant to convey that you take a slice off the index.

jreback replied:
slice is actual slicing; here you are passing a boolean mask that is the same shape (and aligned) with the input); mark this as a separate case

jdeschenes replied:
Hi, in the new case we are not passing a boolean mask. Would you agree if I renamed the test to `test_list_like_indexing`?

jorisvandenbossche replied:
yep, that is better

jorisvandenbossche commented on pandas/tests/indexing/test_timedelta.py:
```diff
@@ -13,7 +13,24 @@ def test_boolean_indexing(self):
                          [10, 10, 10, 3, 4, 5, 6, 7, 8, 9]]
         for cond, data in zip(conditions, expected_data):
-            result = df.assign(x=df.mask(cond, 10).astype('int64'))
+            result = df.assign(x=df.mask(cond, 10).astype(df['x'].dtype))
```
Is this change related to the PR?

jdeschenes replied:
Not at all. The circle CI unit test fails on a 32 bit machine.

jdeschenes replied:
It still fails even with that attempted fix. Travis-CI builds is giving me error entirely unrelated to the change I have done.

jreback commented:
do u have a repo for isnull or is_timeselta64_dtype raising?

jdeschenes commented:
Yeah... I am not sure about this change for the is_timedelta64_dtype. On one hand the behavior to me is bogus(Either a value is of a type or it's not). On the other hand, other projects may rely on this behavior(Why? I have no idea). I ran the test against the pandas suite and nothing broke.

I can make the change locally in get_loc, but I would rather fix the issue directly in the function. Let me know what you think.

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -396,7 +396,10 @@ def is_timedelta64_dtype(arr_or_dtype):
-    tipo = _get_dtype_type(arr_or_dtype)
+    try:
+        tipo = _get_dtype_type(arr_or_dtype)
+    except ValueError:
```
this is fine
but need tests that hit this

Commit:
Changes based on review comments

* Fixed bug on windows with int32/int64 confusion
* Added unit_test to is_timedelta64_dtype when a string is passed
* Changed unit_test name from `test_slice_indexing` to `test_list_like_indexing`

Commit:
* Fixed bug on windows with int32/int64 confusion

jdeschenes commented:
@jreback @jorisvandenbossche Both failures are unrelated to the code I changed. I think it's ready to be merged at this point.

jreback commented on pandas/tests/indexing/test_timedelta.py:
```diff
@@ -16,5 +16,24 @@ def test_boolean_indexing(self):
+        # GH 16637
+        df = pd.DataFrame({'x': range(10)}, dtype="int64")
+        df.index = pd.to_timedelta(range(10), unit='s')
+
```
might be slightly more clear as a parametrized test

e.g.

```
@pytest.mark.parametrize("indexer, exp", [(0, ....), (slice(4, 8), ...), ([3, 5], ...)]
def test_list_like(self, indexer, exp):
    ....
```

jdeschenes replied:
Ok, I am not familiar with this... I will take a look.

jreback commented on pandas/tests/dtypes/test_common.py:
```diff
@@ -204,6 +204,8 @@ def test_is_timedelta64_dtype():
     assert com.is_timedelta64_dtype(np.timedelta64)
     assert com.is_timedelta64_dtype(pd.Series([], dtype="timedelta64[ns]"))
 
+    assert not com.is_timedelta64_dtype("0 days 00:00:00")
```
didn't you have 2 cases that trigger the exception? if so pls add the other one.

jdeschenes replied:
No that I am aware of. There is a separate issue I think were isnull would trigger an exception in the get_loc function. I don't see how that condition could be hit in the future.

jreback commented on doc/source/whatsnew/v0.20.3.txt:
```diff
@@ -37,7 +37,7 @@ Performance Improvements
 ~~~~~~~~~
 
-
+- Bug in ``TimedeltaIndex`` raising a ``ValueError`` when slice indexing with ``loc`` (:issue:`16637`)
```
Bug in a DataFrame/Series with a TimedeltaIndex when slice indexing

jreback commented:
lgtm. pls rebase and push again (had some ci failures unrelated).

jreback commented:
http://pandas.pydata.org/pandas-docs/stable/contributing.html#using-pytest

Commit:
Final change based on @jreback comments

Commit:
Using pytest parametrize as per @jreback suggestion.

jdeschenes commented:
Hi,

I have noticed another issue closely related to this.

```python
df.loc[[0, 3], 'x'] = 20
```

This will fail with the ValueError. Instead, it should do `KeyError: 'None of [[0, 3]] are in the [index]'`

I will take a second look a bit later and see if I can fix this.

jreback commented:
you can address that other error in another PR (or here) up 2 u

jdeschenes commented:
I Will do it in a separate PR

On Jun 9, 2017 7:30 PM, "Jeff Reback" <notifications@github.com> wrote:

> you can address that other error in another PR (or here) up 2 u
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/16638#issuecomment-307523934>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJfOkHsSqoNiRUy1nTRZw-_h-wRaAgUyks5sCdWhgaJpZM4N0GMz>
> .
>

jreback commented:
@jdeschenes ok great. if you can create an issue would be great.

FYI travis is fixed, working on circle now. pytest-xdist new version broke some things.

Commit:
Unneeded try except.

jreback commented on doc/source/whatsnew/v0.20.3.txt:
```diff
@@ -36,11 +36,11 @@ Performance Improvements
 
 Bug Fixes
 ~~~~~~~~~
+- Bug in a DataFrame/Series with a TimedeltaIndex when slice indexing (:issue:`16637`)
```
use double-backticks on TimedeltaIndex

jreback reviewed (changes requested):
if you can add a few tests on the is_timedelta64_dtype (in tests/dtypes/test_common.py)

otherwise lgtm.

TomAugspurger commented:
@jdeschenes can you update? (some linting failures too I believe)

Commit:
Corrections from @jreback comments

* Added test for is_timedelta64_dtype
* Added double backticks in what's new

jdeschenes commented:
I have added the backticks to the what's new.

For the test, I have added an extra case. I can more if you want, if you have a better idea of what you want to see there, let me know.

TomAugspurger commented:
@jdeschenes that's what failed this build: https://travis-ci.org/pandas-dev/pandas/jobs/249668871#L1779

Commit:
Fixed lint issue

jreback commented:
can you rebase

TomAugspurger committed:
Merge branch 'master' into bug_delta_time

jdeschenes commented:
Let me know if you need anything else!

TomAugspurger committed:
Update v0.20.3.txt

Fixed duplicate line in readme

jreback commented:
thanks!
